actix-server/src/worker.rs

```diff
@@ -403,60 +421,61 @@ impl Future for ServerWorker {
 
                 trace!(
                     "Service {:?} has been restarted",
-                    self.factories[idx].name(token)
+                    this.factories[idx].name(token)
                 );
 
-                self.services[token.0].created(service);
-                self.state = WorkerState::Unavailable;
+                this.services[token.0].created(service);
+                this.state = WorkerState::Unavailable;
 
                 self.poll(cx)
             }
-            WorkerState::Shutdown(ref mut t1, ref mut t2, ref mut tx) => {
-                let num = num_connections();
-                if num == 0 {
-                    let _ = tx.take().unwrap().send(true);
+            WorkerState::Shutdown(ref mut timer, ref start_from, _) => {
+                // Wait for 1 second.
+                ready!(timer.as_mut().poll(cx));
+
+                if num_connections() == 0 {
+                    // Graceful shutdown.
+                    if let WorkerState::Shutdown(_, _, sender) = mem::take(&mut this.state) {
```

Comment:
take sender state from outer match ?

Reply:
This is a perf trick to only copy the address when graceful shutdown happen.
Other variant only need to borrow the state rather than take it.

Reply:
It's an extra branch on an irrefutable pattern; the only difference is that this will set state to unavailable too. What's up with replacing the if-let with `sender.send(true)`

Reply:
It's an extra branch. The only difference is that mem::take will also set state back to unavailable but that doesn't change anything. What's up with `sender.send(true)`.

Reply:
This branch only happens when server goes into graceful shutdown so extra branch is not important at all. 

The `mem::take` is to take ownership of `Sender`. oneshot sender take self when calling send API. 
It was an Option<Sender> for the same reason of taking the ownership. But since Sender here is a tokio type so Rust have problem to optimize away the Option wrapper and it can take 8 bytes extra size.

Reply:
oneshot sender takes self... right